Build promotion pipeline YAML

JohnTortugo opened the pull request:
Closes: #4601 

Changes:

- Most of the changes are to plumb parameters fetched from BAR/Maestro to the YAML stages/jobs.
- Other set of changes is to patch DownloadBuildArtifacts tasks to be able to download from the current build or from a pointed build.
- Renamed some files to make more clear the distinction of promote build logic and assign build to channel logic.

I've been testing the Build Promotion [in this pipeline](https://dnceng.visualstudio.com/internal/_build/results?buildId=479697&view=results).

I also tested if the post-build.yml templates are still working when not used for promoting a build. [See test here](https://dnceng.visualstudio.com/internal/_build/results?buildId=479710&view=results).

riarenas commented:
Is there a reason why most of this is in eng/common? There's going to be a single build promotion pipeline that should take the yml from Arcade.

I'd like us to decouple the release variables that are really common from whatever is only needed for channel assignment after the fact. The pipeline can still reference the stages in eng/common, but unless there's something I'm missing, I don't think we need to distribute all of these changes to all repos.

JohnTortugo commented:
@riarenas - I addressed the comments we discussed offline: no new files in eng\common

Note that some variables are used only in build promotion and others only at build creation time. Other variables are used in both cases. The major change here is because we need to be able to download artifacts from the current build or from another [completed] build.

I use the same Job in `setup-maestro-vars.yml` for all vars because that reduces the number of changes to the other files. Otherwise I'd have to add a `coalesce(xxxx, yyy)` in every place we refer to a variable declared in that Job.

Testing the build promotion: https://dnceng.visualstudio.com/internal/_build/results?buildId=488127&view=results
Testing the post-build YAMLs for build creation: https://dnceng.visualstudio.com/internal/_build/results?buildId=488352&view=results
Testing the post-build YAMLs for build creation + activating a channel: https://dnceng.visualstudio.com/internal/_build/results?buildId=488415&view=results

JohnTortugo commented:
Ping

riarenas commented:
Sorry I missed the latest update. Looking now.

riarenas commented on eng/common/templates/post-build/setup-maestro-vars.yml:
```diff
@@ -19,22 +26,61 @@ jobs:
         script: |
           try {
-            $Content = Get-Content $(Build.StagingDirectory)/ReleaseConfigs/ReleaseConfigs.txt
+            if ($Env:PromoteToMaestroChannelId -eq 0) {
```
I'd just add a note to this block saying we want to keep it as an inline script to avoid checking out the entire repo just for this job.

riarenas commented on eng/common/templates/post-build/setup-maestro-vars.yml:
```diff
@@ -19,22 +26,61 @@ jobs:
             exit 1
           }
+      env:
+        MAESTRO_API_TOKEN: $(MaestroApiAccessToken)
```
Where is the value for this coming from? is it from some other YAML file that imports a variable group that has this secret?

riarenas replied:
Not sure if that is done through YAML, or in the actual pipeline.

JohnTortugo replied:
It comes from the variables template imported on line 5: common-variables.yml

riarenas replied:
Got it. Makes sense, knew it had to be coming from somewhere.

riarenas commented on eng/promote-build.yml:
```diff
@@ -0,0 +1,35 @@
+#
+#   - BARBuildId
+#   - PromoteToMaestroChannelId
+#   - EnableSymbolValidation
```
Can we add default values to most of these?

riarenas replied:
Not sure if that is done through YAML, or in the actual pipeline.

JohnTortugo replied:
It's done in the pipeline. I can't add the variables here or we won't be able to set them at queue time.

riarenas replied:
Got it. A little annoying but I understand.

riarenas commented on eng/promote-build.yml:
```diff
@@ -0,0 +1,35 @@
+#   - EnableSourceLinkValidation
+#   - enableNugetValidation
+#   - EnableSigningValidation
+#   - EnableSDLValidation
```
Hmm, I wonder if there's any value to enabling SDL in this pipeline. Would anything have changed between the original build and this one? we'd be examining the same source and produced bits, and this parameter seems like it will be annoying to fill out during queue time?

riarenas replied:
I think the question applies to everything in the validate stage.

JohnTortugo replied:
That's a good point. My idea here was that since this pipeline should be generic it could be used for promoting a build that was created with SDL not enabled.

Perhaps just setting the parameter to false by default would do the trick?

riarenas replied:
Yeah, probably enough to have these as false by default, except maybe the sdl additional parameters still seems super hard to fill out. Have you tried passing it during queue time? does it look reasonable for someone to be able to fill that out?

JohnTortugo replied:
It's hard to fill out and it's hard to set a generic (default) value for it; it uses the repository name for instance.

riarenas commented:
The test build at https://dnceng.visualstudio.com/internal/_build/results?buildId=479710&view=results failed. Maybe not caused by the changes, but the errors happened in one of the publishing tasks, so I'd like to have a successful build + publish with these changes before merge.

JohnTortugo commented:
Re-run again the tests:

- [Build promotion pipeline](https://dnceng.visualstudio.com/internal/_build/results?buildId=495381&view=results)
- [Build creation pipeline](https://dnceng.visualstudio.com/internal/_build/results?buildId=495427&view=results)

riarenas commented:
> Build promotion pipeline

This one wasn't set up to publish anywhere, so all jobs were skipped. We had already seen this succeed though, but if you want to make double sure, set up a default channel 😁

JohnTortugo commented:
Tested it again here, now publishing to some channel: 
- https://dnceng.visualstudio.com/internal/_build/results?buildId=496071&view=results
- https://dnceng.visualstudio.com/internal/_build/results?buildId=496027&view=results

Also added the check to validate that the channel exist.

riarenas commented:
Looks good to me, but I started thinking that it's not easy to tell what the pipeline is doing just from looking at the build.

I think we can look at what Davis did in this PR https://github.com/dotnet/arcade/pull/4664/files so that when looking at a build for that pipeline, it's clear which build was promoted, and to which channel. What do you think?

JohnTortugo commented:
> Looks good to me, but I started thinking that it's not easy to tell what the pipeline is doing just from looking at the build.
> 
> I think we can look at what Davis did in this PR https://github.com/dotnet/arcade/pull/4664/files so that when looking at a build for that pipeline, it's clear which build was promoted, and to which channel. What do you think?

Yeah, looks simple and useful enough to include in this PR.

JohnTortugo commented:
Test build with build tags: https://dnceng.visualstudio.com/internal/_build/results?buildId=496298&view=results

riarenas commented on eng/common/templates/post-build/setup-maestro-vars.yml:
```diff
@@ -71,6 +71,28 @@ jobs:
                 exit 1
               }
 
+              $azureDevOpsBuildNumber = $buildInfo.azureDevOpsBuildNumber
```
Does this make more sense to be in the promote-build yaml instead of in the common file?

JohnTortugo commented:
New test build here: https://dnceng.visualstudio.com/internal/_build/results?buildId=496588&view=results

JohnTortugo commented:
Fixed a bug while enabling/disabling conditions. Ran new tests to test that part specifically:

- [build promotion](https://dnceng.visualstudio.com/internal/_build/results?buildId=497218&view=results)
- [build creation](https://dnceng.visualstudio.com/internal/_build/results?buildId=497543&view=results)

tmat commented on eng/promote-build.yml:
```diff
@@ -0,0 +1,104 @@
+          inputs:
+            targetType: inline
+            script: |
+              # Keeping this script inline so that we don't need to checkout the whole repo to use just one file
```
A while ago I suggested moving scripts that are used in post-build phases into a NuGet package, so that the build pipeline can restore just that one nuget package and have all the scripts that it needs available. If it was done that way you wouldn't need to restore an entire repo just to get the scripts, or inline the script into YAML.

JohnTortugo replied:
Thanks @tmat . I created an issue to discuss about this: https://github.com/dotnet/arcade/issues/4755
